File: 148. Sort List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def sortList(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        slow = fast = head
        prev = None
        while fast and fast.next:
            prev = slow
            fast = fast.next.next
            slow = slow.next
        prev.next = None
        left = self.sortList(head)
        right = self.sortList(slow)
        head = self.merge(left, right)
        return head
    
    def merge(self, left, right):
        isLeft = 0
        if left.val < right.val:
            head = curr = left
            left = left.next
            isLeft = 1
        else:
            head = curr = right
            right = right.next
        while left and right:
            if left.val < right.val:
                curr.next = left
                left = left.next
                curr = curr.next
                isLeft = 1
            else:
                curr.next = right
                right = right.next
                curr = curr.next
                isLeft = 0
        if isLeft == 0:
            curr.next = left
        else:
            curr.next = right
        return head

File: 148. Sort List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode fast=head,slow=head,prev=null;
        while (fast != null && fast.next != null) {
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }
        prev.next = null;
        ListNode h1 = sortList(head);
        ListNode h2 = sortList(slow);
        if (h1 == null) {
            return h2;
        } else if (h2 == null) {
            return h1;
        }
        ListNode newhead;
        if (h1.val < h2.val) {
            newhead = h1;
            h1 = h1.next;
        } else {
            newhead = h2;
            h2 = h2.next;
        }
        ListNode node = newhead;
        while (h1 != null && h2 != null) {
            if (h1.val < h2.val) {
                node.next = h1;
                h1 = h1.next;
            } else {
                node.next = h2;
                h2 = h2.next;
            }
            node = node.next;
        }
        if (h1 == null) {
            node.next = h2;
        } else if (h2 == null) {
            node.next = h1;
        }
        return newhead;
    }
}